aspnetcore/host-and-deploy/health-checks.md

```diff
@@ -0,0 +1,359 @@
+
+```console
+dotnet run --scenario basic
+```
+
+### Docker example
+
+[Docker](xref:host-and-deploy/docker/index) offers a built-in `HEALTHCHECK` directive that can be used to check the status of an app that uses the basic health check configuration:
+
+```
+HEALTHCHECK CMD curl --fail http://localhost:5000/health || exit
+```
+
+## Create health checks
+
+Health checks are created by implementing the `IHealthCheck` interface. The `IHealthCheck.CheckHealthAsync` method returns a `Task<HealthCheckResult>` of either `Healthy`, `Unhealthy`, `Degraded`, or `Failed`. The result is written as a plain text response with a configurable status code (configuration is described in the [Health check options](#health-check-options) section).
+
+The following `ExampleHealthCheck` class demonstrates the layout of a health check:
+
+```csharp
+public class ExampleHealthCheck : IHealthCheck
+{
+    public ExampleHealthCheck()
+    {
+        // Use DI to supply any required services to the health check.
+    }
+
+    public string Name => "example_check";
+
+    public Task<HealthCheckResult> CheckHealthAsync(
+        CancellationToken cancellationToken = default(CancellationToken))
+    {
+        // Execute health check logic here. This example sets a dummy
+        // variable to true.
+        var healthCheckResultHealthy = true;
+
+        // Return a HealthCheckResult value depending on the outcome of the
+        // check(s). Other HealthCheckResult outcomes not shown in this
+        // example are Degraded and Failed.
+        if (healthCheckResultHealthy)
+        {
+            return Task.FromResult(
+                HealthCheckResult.Healthy("The check indicates a healthy state."));
+        }
+
+        return Task.FromResult(
+            HealthCheckResult.Unhealthy("The check indicates an unhealthy state."));
+    }
+}
+```
+
+Add the `ExampleHealthCheck` with `AddCheck` and call `UseHealthChecks` in the processing pipeline with the endpoint URL:
+
+```csharp
+public void ConfigureServices(IServiceCollection services)
+{
+    services.AddHealthChecks()
+        .AddCheck<ExampleHealthCheck>();
+}
+
+public void Configure(IApplicationBuilder app, IHostingEnvironment env)
+{
+    app.UseHealthChecks("/health");
+}
+```
+
+`AddCheck` can also execute a lambda function. In the following example, the health check name is specified as `Foo` and the check always returns a healthy state:
+
+```csharp
+public void ConfigureServices(IServiceCollection services)
+{
+    services.AddHealthChecks()
+        .AddCheck("Foo", () => Task.FromResult(
+            HealthCheckResult.Healthy("Foo is OK!")));
+}
+```
+
+## Health check options
+
+`HealthCheckOptions` provide an opportunity to customize health check behavior:
+
+* [Filter health checks](#filter-health-checks)
+* [Customize the HTTP status code](#customize-the-http-status-code)
```

Comment:
> suppress the anti-caching headers

The language logic of `SuppressCacheHeaders` seems harder to grok than needed. Did you reject options that pertain to the result of setting the headers and not what the property does to the headers themselves ...

* `SuppressCachedResponses`, `PreventCachedResponses`, `SuppressCaching` or `PreventCaching`
   * `true` adds the cache-suppression headers.
   * `false` leaves caching headers alone.
* `AllowCachedResponses` or `CacheResponses`
   * `true` leaves caching headers alone.
   * `false` adds the cache-suppression headers.

Sure, I'll put what `true`/`false` mean as it stands; but when I read the phrase "suppress cache headers," it's not immediately evident what's going to happen when I set the value one way or ther other. Actually when I see the property name, I think it's going to do the exact opposite of what it does ... I think it's going to suppress whatever caching headers the **_app already has set_** and apply a set that will stop caching.

Reply:
Yeah I think that's fair feedback. I'll log an issue and we'll consider changing it. In general any boolean options are phrased the way they are so that it makes sense when `false` is the default.

I think the best option name I've heard so far is `AllowCachedResponses` or `AllowCachingResponses`.